lib/tcpip/network_wrapper.cc

```diff
@@ -590,6 +689,128 @@ SObj network_socket_create_and_bind(Timeout       *timeout,
 	  static_cast<SObj>(nullptr) /* return nullptr if we are restarting */);
 }
 
+SObj network_socket_accept_tcp(Timeout        *timeout,
+                               SObj            mallocCapability,
+                               SObj            sealedListeningSocket,
+                               NetworkAddress *address,
+                               uint16_t       *port)
+{
+	SObj socket = nullptr;
+	with_sealed_socket(
+	  [&](SealedSocket *listeningSocket) {
+		  if (!check_timeout_pointer(timeout))
+		  {
+			  return -EINVAL;
+		  }
+
+		  // Create a sealed socket wrapper for the accepted connection.
+		  auto [socketWrapper, sealedSocket] = token_allocate<SealedSocket>(
+		    timeout, mallocCapability, socket_key());
+		  if (socketWrapper == nullptr)
+		  {
+			  Debug::log("Failed to allocate socket wrapper.");
+			  return -EINVAL;
+		  }
+
+		  socketWrapper->socketEpoch = currentSocketEpoch.load();
+
+		  struct freertos_sockaddr addressTmp;
+		  uint32_t                 addressLength = sizeof(addressTmp);
+		  auto                     rawSocket     = FreeRTOS_accept(
+		                            listeningSocket->socket, &addressTmp, &addressLength);
+		  if (rawSocket == nullptr)
+		  {
+			  Debug::log("Failed to create socket.");
+			  // This cannot fail unless buggy - we know that we
+			  // successfully allocated the token with this malloc
+			  // capability. Same for other calls to `token_obj_destroy`
+			  // in this function.
+			  token_obj_destroy(mallocCapability, socket_key(), sealedSocket);
+			  return -EINVAL;
+		  }
+		  socketWrapper->socket = rawSocket;
+
+		  // Claim the socket so that it counts towards the caller's quota.  The
```

Comment:
I think this is copied from my code, but I think it's wrong and should be fixed in both places.

If we allocate with the network stack's capability and claim with the caller's capability, then we are the owner.  We cannot free with heap_free_all because it needs the sealing key, but they can call heap_free_all to drop their claim.  You can use this to exhaust the network stack's quota.

Reply:
David and I discussed this offline, and it sounds like this is a pre-existing, larger problem. I described the issue in https://github.com/CHERIoT-Platform/cheriot-rtos/issues/140 and we will address this separately.